File: scripts/generate-snapshot.py
#!/usr/bin/env python3
"""
Generate specs/SNAPSHOT.md from all active spec files.

Usage: python generate-snapshot.py --specs-dir specs/
"""

import argparse
import re
from pathlib import Path
from datetime import datetime
from collections import defaultdict


def parse_frontmatter(content: str) -> dict:
    """Extract YAML frontmatter from markdown."""
    match = re.match(r'^---\s*\n(.*?)\n---', content, re.DOTALL)
    if not match:
        return {}

    frontmatter = {}
    for line in match.group(1).split('\n'):
        if ':' in line:
            key, value = line.split(':', 1)
            frontmatter[key.strip()] = value.strip()
    return frontmatter


def extract_overview(content: str) -> str:
    """Extract the overview section from spec content."""
    # Remove frontmatter
    content = re.sub(r'^---\s*\n.*?\n---\s*\n', '', content, flags=re.DOTALL)

    # Find overview section
    match = re.search(r'## Overview\s*\n(.*?)(?=\n##|$)', content, re.DOTALL)
    if match:
        return match.group(1).strip()
    return ""


def generate_snapshot(specs_dir: Path) -> str:
    """Generate SNAPSHOT.md content."""
    # Find all active spec files
    specs = []
    for spec_path in specs_dir.rglob('*.md'):
        if spec_path.name in ('INDEX.md', 'SNAPSHOT.md', 'glossary.md'):
            continue

        content = spec_path.read_text()
        fm = parse_frontmatter(content)

        if fm.get('status') == 'active':
            specs.append({
                'title': fm.get('title', spec_path.stem),
                'path': str(spec_path.relative_to(specs_dir)),
                'domain': fm.get('domain', 'Unknown'),
                'issue': fm.get('issue', ''),
                'overview': extract_overview(content),
                'content': content,
            })

    # Group by domain
    by_domain = defaultdict(list)
    for spec in specs:
        by_domain[spec['domain']].append(spec)

    # Generate markdown
    lines = [
        "# Product Snapshot",
        "",
        f"Generated: {datetime.now().strftime('%Y-%m-%d')}",
        "",
        "This document represents the current active state of the product by compiling all active specifications.",
        "",
    ]

    # Table of contents
    if by_domain:
        lines.extend([
            "## Table of Contents",
            "",
        ])
        for domain in sorted(by_domain.keys()):
            lines.append(f"- [{domain}](#{domain.lower().replace(' ', '-')})")
        lines.append("")

    # By domain
    lines.extend([
        "## By Domain",
        "",
    ])

    for domain in sorted(by_domain.keys()):
        lines.extend([
            f"### {domain}",
            "",
        ])

        for spec in sorted(by_domain[domain], key=lambda x: x['title']):
            lines.extend([
                f"#### {spec['title']}",
                f"**Spec:** [{spec['path']}]({spec['path']})",
            ])

            if spec['issue']:
                lines.append(f"**Issue:** [{spec['issue']}](#)")

            lines.append("")

            if spec['overview']:
                lines.append(spec['overview'])
                lines.append("")

            lines.append("---")
            lines.append("")

    if not by_domain:
        lines.append("*No active specs yet*")
        lines.append("")

    return '\n'.join(lines)


def main():
    parser = argparse.ArgumentParser(description='Generate product snapshot')
    parser.add_argument('--specs-dir', default='specs/', help='Specs directory')
    args = parser.parse_args()

    specs_dir = Path(args.specs_dir)
    if not specs_dir.exists():
        print(f"Error: Specs directory not found: {specs_dir}")
        return 1

    snapshot_content = generate_snapshot(specs_dir)
    snapshot_path = specs_dir / 'SNAPSHOT.md'
    snapshot_path.write_text(snapshot_content)

    print(f"✓ Generated {snapshot_path}")
    return 0


if __name__ == '__main__':
    import sys
    sys.exit(main())

File: plugins/sdd/scripts/generate-snapshot.ts
#!/usr/bin/env npx ts-node --esm
/**
 * Generate specs/SNAPSHOT.md from all active spec files.
 *
 * Usage: npx ts-node --esm generate-snapshot.ts --specs-dir specs/
 */

import * as fs from 'node:fs/promises';
import * as path from 'node:path';
import { extractOverview } from './lib/frontmatter.js';
import { findSpecFiles, directoryExists } from './lib/spec-utils.js';

interface ActiveSpec {
  readonly title: string;
  readonly path: string;
  readonly domain: string;
  readonly issue: string;
  readonly overview: string;
}

/**
 * Generate SNAPSHOT.md content.
 */
const generateSnapshot = async (specsDir: string): Promise<string> => {
  const allSpecs = await findSpecFiles(specsDir);

  // Filter to active specs only
  const specs: ActiveSpec[] = [];
  for (const spec of allSpecs) {
    const fm = spec.frontmatter ?? {};
    if (fm['status'] === 'active') {
      specs.push({
        title: fm['title'] ?? path.basename(spec.path, '.md'),
        path: spec.relativePath,
        domain: fm['domain'] ?? 'Unknown',
        issue: fm['issue'] ?? '',
        overview: extractOverview(spec.content),
      });
    }
  }

  // Group by domain
  const byDomain: Record<string, ActiveSpec[]> = {};
  for (const spec of specs) {
    if (!byDomain[spec.domain]) {
      byDomain[spec.domain] = [];
    }
    byDomain[spec.domain]?.push(spec);
  }

  const today = new Date().toISOString().split('T')[0];
  const domains = Object.keys(byDomain).sort();

  // Generate markdown
  const lines: string[] = [
    '# Product Snapshot',
    '',
    `Generated: ${today}`,
    '',
    'This document represents the current active state of the product by compiling all active specifications.',
    '',
  ];

  // Table of contents
  if (domains.length > 0) {
    lines.push('## Table of Contents');
    lines.push('');
    for (const domain of domains) {
      const anchor = domain.toLowerCase().replace(/ /g, '-');
      lines.push(`- [${domain}](#${anchor})`);
    }
    lines.push('');
  }

  // By domain
  lines.push('## By Domain');
  lines.push('');

  for (const domain of domains) {
    lines.push(`### ${domain}`);
    lines.push('');

    const domainSpecs = byDomain[domain] ?? [];
    const sorted = [...domainSpecs].sort((a, b) => a.title.localeCompare(b.title));

    for (const spec of sorted) {
      lines.push(`#### ${spec.title}`);
      lines.push(`**Spec:** [${spec.path}](${spec.path})`);

      if (spec.issue) {
        lines.push(`**Issue:** [${spec.issue}](#)`);
      }

      lines.push('');

      if (spec.overview) {
        lines.push(spec.overview);
        lines.push('');
      }

      lines.push('---');
      lines.push('');
    }
  }

  if (domains.length === 0) {
    lines.push('*No active specs yet*');
    lines.push('');
  }

  return lines.join('\n');
};

/**
 * Parse command line arguments.
 */
const parseArgs = (args: readonly string[]): { specsDir: string } => {
  let specsDir = 'specs/';

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === '--specs-dir') {
      specsDir = args[i + 1] ?? 'specs/';
      i++;
    }
  }

  return { specsDir };
};

const main = async (): Promise<number> => {
  const args = parseArgs(process.argv.slice(2));

  if (!(await directoryExists(args.specsDir))) {
    console.error(`Error: Specs directory not found: ${args.specsDir}`);
    return 1;
  }

  const snapshotContent = await generateSnapshot(args.specsDir);
  const snapshotPath = path.join(args.specsDir, 'SNAPSHOT.md');
  await fs.writeFile(snapshotPath, snapshotContent);

  console.log(`✓ Generated ${snapshotPath}`);
  return 0;
};

main()
  .then(process.exit)
  .catch((err) => {
    console.error(err);
    process.exit(1);
  });
